src/DoubleSlider.elm

```diff
@@ -199,114 +102,186 @@ onInsideRangeClick model =
+inputDecoder (DoubleSlider slider) thumb =
+    Json.Decode.map (\value -> String.toFloat value |> Maybe.withDefault 0 |> convertValue (DoubleSlider slider) thumb)
+        Html.Events.targetValue
+
 
+convertValue : DoubleSlider msg -> Thumb -> Float -> Float
+convertValue (DoubleSlider slider) thumb value =
+    case thumb of
+        Low ->
+            Basics.min value (slider.highValueAttributes.value - (slider.commonAttributes.step * slider.overlapThreshold))
 
-{-| DEPRECATED: Displays the slider
--}
-fallbackView : Model -> Html Msg
-fallbackView model =
-    view model
+        High ->
+            Basics.max value (slider.lowValueAttributes.value + (slider.commonAttributes.step * slider.overlapThreshold))
 
 
-{-| Renders the current values using the formatter
--}
-formatCurrentRange : Model -> String
-formatCurrentRange model =
-    if model.lowValue == model.min && model.highValue == model.max then
+
+-- API
+
+
+defaultCurrentRangeFormatter : { lowValue : Float, highValue : Float, min : Float, max : Float } -> String
+defaultCurrentRangeFormatter values =
+    if values.lowValue == values.min && values.highValue == values.max then
         ""
 
     else
-        model.currentRangeFormatter model.lowValue model.highValue model.min model.max
+        String.join " " [ String.fromFloat values.lowValue, "-", String.fromFloat values.highValue ]
 
 
-{-| Returns the subscriptions necessary to run
--}
-subscriptions : Model -> Sub Msg
-subscriptions model =
-    Sub.none
+init :
+    { min : Float
+    , max : Float
+    , step : Float
+    , lowValue : Float
+    , highValue : Float
+    , onLowChange : Float -> msg
+    , onHighChange : Float -> msg
+    , valueFormatter : { value : Float, max : Float } -> String
+    , minFormatter : { value : Float } -> String
+    , maxFormatter : { value : Float } -> String
+    , currentRangeFormatter : { lowValue : Float, highValue : Float, min : Float, max : Float } -> String
+    , overlapThreshold : Float
+    }
+    -> DoubleSlider msg
+init attrs =
+    DoubleSlider
+        { commonAttributes =
+            { min = attrs.min
+            , max = attrs.max
+            , step = attrs.step
+            , minFormatter = attrs.minFormatter
+            , maxFormatter = attrs.maxFormatter
+            }
+        , lowValueAttributes =
+            { value = attrs.lowValue
+            , change = attrs.onLowChange
+            , formatter = attrs.valueFormatter
+            }
+        , highValueAttributes =
+            { value = attrs.highValue
+            , change = attrs.onHighChange
+            , formatter = attrs.valueFormatter
+            }
+        , currentRangeFormatter = attrs.currentRangeFormatter
+        , overlapThreshold = attrs.overlapThreshold
+        }
+
+
+update : { lowValue : Maybe Float, highValue : Maybe Float } -> DoubleSlider msg -> DoubleSlider msg
+update values (DoubleSlider ({ lowValueAttributes, highValueAttributes } as slider)) =
```

Comment:
This could be split into two functions. I guess the common case is to update only one value at a time.

Reply:
💡